### DISCUSSION TITLE
How to match Protocol with a getter-like descriptor?

### DISCUSSION THREAD

**Discussion by GreyElaina** [microsoft/pyright#5970] [Q&A]
as below, the protocol:

```python
class SupportsCollect(Protocol[R]):
    collect: R
```

and if we have a instance but `x.collect` is a descriptor:

```python
class A:
    collector = EgDescriptor()
```

then have a simple function:

```python
def a(x: SupportsCollect[R]):
    return x.collect
```

pass `a` with `A`'s instance, then see that we got...*the descriptor*?

```python
reveal_type(a(A()))  # got `EgDescriptor`
```

i tried simulate the `getattr` behavior via `Protocol[R].__get__`, it donot works, and got `"collect" is invariant because it is mutable`, what can i do for this?

**Comment by erictraut** [comment]
I'm not sure I fully understand the problem you're trying to solve. Your code snippets above reference a bunch of symbols that are undefined. Could you provide a complete, self-contained code sample?

**Reply by erictraut** [reply]
One thought is that descriptor objects are typically designed to work only when they are class variables, so perhaps you need to use `typing.ClassVar[R]` within your protocol definition.

**Reply by GreyElaina** [reply]
i tried but classvar cannot include typevar and so it could not infer variables.
i will fill these missings with a well-wrote snippet.

**Reply by GreyElaina** [reply]
```python
from __future__ import annotations
from typing import Protocol, TypeVar, overload, Any
from typing_extensions import Self

T = TypeVar("T")
R = TypeVar("R", covariant=True)

class SupportsCollect(Protocol[T]):
    collect: T

class Eg:
    @property
    def collect(self):
        ...

def a(x: SupportsCollect[T]) -> T:
    return x.collect

reveal_type(a(Eg()))  # Type of "a(Eg())" is "property"

# "property" is not the only case.

class Getter:
    @overload
    def __get__(self, instance: None, owner: type) -> Self:
        ...
    
    @overload
    def __get__(self, instance: Any, owner: type) -> int:
        ...
    
    def __get__(self, instance: Any, owner: type) -> Self | int:
        if instance is None:
            return self
        
        return 1

class Eg2:
    collect = Getter()

reveal_type(a(Eg2()))  # Type of "a(Eg2())" is "Getter"

# i would like pyright to follow the "getattr" behavior,
# but i encountered this problem, could you please give me some advises?
```

**Reply by GreyElaina** [reply]
btw, `R` in my previous code is "infer_variable=True", sorry.

**Reply by GreyElaina** [reply]
```python
# ClassVar? but pyright said no way.

class SupportsCollect2(Protocol[T]):
    collect: ClassVar[T]  # "ClassVar" type cannot include type variables
```

**Reply by GreyElaina** [reply]
```python
# what about a "Getter" protocol?

class GetterProtocol(Protocol[R]):
    @overload
    def __get__(self, instance: None, owner: type) -> Self:
        ...
    
    @overload
    def __get__(self, instance: Any, owner: type) -> R:
        ...

class SupportsCollect3(Protocol[T]):
    collect: GetterProtocol[T]

def b(x: SupportsCollect3[T]) -> T:
    return x.collect

# in first case, pyright said:
#   Argument of type "Eg" cannot be assigned to parameter "x" of type "SupportsCollect3[T@b]" in function "b"
#     "Eg" is incompatible with protocol "SupportsCollect3[T@b]"
#       "collect" is invariant because it is mutable
#       "collect" is an incompatible type
#         "property" is incompatible with protocol "GetterProtocol[T@SupportsCollect3]"

reveal_type(b(Eg()))

# in second case, "Getter" is compatible with protocol "GetterProtocol", even i wrote in a almost same pattern.

reveal_type(b(Eg2()))

# ...and now, i have no idea to do this.
```

**Reply by erictraut** [reply]
You would need some way of explaining that the protocol describes "a class that has an immutable class variable that contains a descriptor object that returns a generic type".

If you're willing to stick to just properties, the following works:

```python
from __future__ import annotations
from typing import Protocol, TypeVar, overload, Any
from typing_extensions import Self

T = TypeVar("T")
R = TypeVar("R", covariant=True)

class SupportsCollect(Protocol[R]):
    @property
    def collect(self) -> R:
        ...

class Eg:
    @property
    def collect(self) -> int:
        ...

def a(x: SupportsCollect[T]) -> T:
    return x.collect

reveal_type(a(Eg()))
```

(Note: the above also works in mypy.)

If you want to support your own "read only descriptor", this works in pyright:

```python
from __future__ import annotations
from typing import Final, Protocol, TypeVar, overload, Any
from typing_extensions import Self

T = TypeVar("T")
R = TypeVar("R", covariant=True)

class ReadOnlyDescriptor(Protocol[R]):
    @overload
    def __get__(self, instance: None, owner: type) -> Self:
        ...

    @overload
    def __get__(self, instance: Any, owner: type) -> R:
        ...

class SupportsCollect(Protocol[R]):
    collect: Final[ReadOnlyDescriptor[R]]

def a(x: SupportsCollect[T]) -> T:
    return x.collect

class Getter:
    @overload
    def __get__(self, instance: None, owner: type) -> Self:
        ...

    @overload
    def __get__(self, instance: Any, owner: type) -> int:
        ...

    def __get__(self, instance: Any, owner: type) -> Self | int:
        if instance is None:
            return self

        return 1

class Eg2:
    collect: Final[Getter] = Getter()

reveal_type(a(Eg2()))
```

(Note: the second example doesn't work in mypy because it enforces that protocol members can't be final and must be initialized with a value.)

**Reply by GreyElaina** [reply]
seems a working way, i will respond as soon as possible

**Reply by GreyElaina** [reply]
hmmm, pyright also enforces that must be initialized with a value.

```python
class SupportsCollect(Protocol[R]):
    collect: Final[Getter[R]]  # "collect" is declared Final, but value is not assigned
```

**Reply by erictraut** [reply]
> hmmm, pyright also enforces that must be initialized with a value.

I changed this recently. If you upgrade to pyright 1.1.327 (last week's release), you'll see that it no longer enforces this.

**Reply by GreyElaina** [reply]
i will do some experiments, after that i will summarize a solution for who encounter this problem like me.

heart for your great work and support!

**Comment by Azureblade3808** [comment]
Sample for an ugly work-around --

```python
from typing import Any, Generic, Protocol, TypeVar

R_co = TypeVar("R_co", covariant=True)

class DescriptorProtocol(Protocol[R_co]):
    def __get__(self, instance: Any, owner: type[Any] | None, /) -> R_co:
        ...

class Descriptor(Generic[R_co]):
    def __get__(self, instance: Any, owner: type[Any] | None, /) -> R_co:
        raise NotImplementedError

class P(Protocol[R_co]):
    # Using `property` to show this attribute is read-only.
    @property
    def r(self, /) -> DescriptorProtocol[R_co]:
        ...

class C:
    r = Descriptor[int]()  # Can be defined with `Final` or `ClassVar`.

x: P[int] = C()  # Passes type-checks.
```

The code may seem weird but it currently works for both Pyright and Mypy.

**Reply by Azureblade3808** [reply]
Sorry, this one isn't working. It just fails for a plain `@property`.

**Comment by GreyElaina** [comment]
i found another method to solve my problem (via a big refactor :(

might i can close even it's not solved?

**Comment by GreyElaina** [comment]
[code snippet](https://pyright-playground.decorator-factory.su/?gzip=H4sIADHwCmUC_32SQU_DMAyF7_kV1k6pVCrtOmkIqDSuXZm4TKjyOhciZUmVpmP99zijYd1A5Oi89zl-TuPsAaqq6X3vqKpAHVrrPKAx1qNX1nRCNEHjh1aZ93j_aIYUctQad5pSeCZDTtUpFM56W1udwmZo6RXdlbuikyfTBWoEFejw8NISex0dCXXFUhJiA8uIkLPNLBEFF37EclZwqZxqylkKtT2iU2j8cuN6YsH8SjH_QyLyqSRnxc72Zr-8zHbryNdXjvWfFuMd3jSqNXYdrIwcw9rmHNJbshDAZ08Nb0EZ5atKdqSbFHAB-XgbTihmyK0xol76NkTY5VZrqr2M2W_zdQrllHwg_2H3I_e04EecNbzEtwTu7qG8tMmyLPKf6AOPyrr_OePQ2yL0HJkXXHB9OmxbcnJYwLV6ogvHEf9BA4O4KYx-IQIMJfe9mfzXI8p5ZI-IU_b9dCGU6TyamjjIOJ_k5TyszBm_k8jr5A2ec-n8OPw5lcn_lCgjKJG7JBFfWRCCeEcDAAA%3D)

```
Type "(a: Unknown, b: Unknown, c: Unknown) -> str" cannot be assigned to type "(**P@method) -> R@method"
        Function return type "str" is incompatible with type "R@method"
          Type "str" cannot be assigned to type "R@method"
```

```
Type of "a(instance)(b)" is "(y: (**P@method) -> R@method) -> ((**P@method) -> R@method)"
```

Strange.....

**Reply by GreyElaina** [reply]
why not "infer" `R` as `str`...?

**Comment by draustin** [comment]
Is there a way of explaining that a Protocol has an instance attribute that is matched by a descriptor as well as a regular instance attribute? The following explains what I'd like to achieve for the case of a read-only instance attribute.

```python
R = TypeVar("R", covariant=True)


class ReadOnlyDescriptorProtcol(Protocol[R]):
    def __get__(self, obj: Any, typ: type) -> R:
        ...
    
    # Tried adding the following too.
    # @overload
    # def __get__(self, obj: None, typ: type) -> Self:
    #    ...

class HasReadOnlyX(Protocol):
    # Trying to say that instances of `HasReadOnlyX` have a read-only attribute `x`, which is an int.
    x: ReadOnlyDescriptorProtcol[int]


class AlsoHasX:
    def __init__(self):
        self.x = 1


# Works with any object that has an int attribute 'x'.
def f(arg: HasReadOnlyX):
    print(arg.x + 1)


# An instance of B has an int attribute x, so the following seems OK to me:
f(AlsoHasX())
# But Pyright strict mode says:
#
# Argument of type "B" cannot be assigned to parameter "arg" of type "HasReadOnlyX" in function "f"
#   "B" is incompatible with protocol "HasReadOnlyX"
#     "x" is invariant because it is mutable
#     "x" is an incompatible type
#       "int" is incompatible with protocol "ReadOnlyDescriptorProtcol[int]"PylancereportGeneralTypeIssues

```

I get that there are subtle differences between an `AlsoHasX` and an implementation that actually uses a descriptor, but *think* that these differences only arise through introspection or accessing the class object. So if one sticks to normal attribute access (i.e. `arg.x` in `f` above) then passing an `AlsoHasX` to `f` is type safe. Is there any way of explaining this to Pyright? Many thanks.

**Reply by erictraut** [reply]
According to PEP 544:

> To define a read-only protocol variable, one can use an (abstract) property.

Pyright accordingly special-cases `@property` within protocols.

```python
class HasReadOnlyX(Protocol):
    @property
    def x(self) -> int:
        ...
```

It does this in a manner that is type safe even when accessing the class object.

**Reply by draustin** [reply]
Thanks very much. Allow me to give a little more context. I'm not using regular `@property` because it is not generic and I need to access its type elsewhere in the class definition. I am building a framework for signals (similar to the Qt Signal/Slot system) and would like to be able to declare a signal representing the change of a property. Ideally I'd do something like the following, except that `property` isn't generic.

```python

R = TypeVar("R")

# Not allowed because property isn't generic.
def observer(p: property[R]) -> ChangeSignalDescriptor[R]:
    # Access p.fget and construct signal object.
    ...

class HasReadOnlyX(Protocol):
    @property
    def x(self) -> int:
        ...
        
    x_changed = observer(x)

```
The main purpose of rolling my own property descriptor is so it can be generic.

**Reply by erictraut** [reply]
This isn't as elegant as I'd like, but I think it meets your requirements (at least as I understand them). Unfortunately, I couldn't find a way to accommodate all of your requirements in mypy, so this solution works only in pyright.

```python
from typing import Any, Final, Generic, Protocol, TypeVar

R = TypeVar("R", covariant=True)

class ReadOnlyDescriptorProtocol(Protocol[R]):
    def __get__(self, obj: Any, typ: type) -> R:
        ...

class HasReadOnlyXDescriptorProtocol(Protocol[R]):
    x: Final[ReadOnlyDescriptorProtocol[R]]

class HasXPropertyProtocol(Protocol[R]):
    @property
    def x(self) -> R:
        ...

HasReadOnlyXProtocol = HasReadOnlyXDescriptorProtocol[R] | HasXPropertyProtocol[R]

class ReadOnlyDescriptor(Generic[R]):
    def __get__(self, obj: Any, typ: Any) -> R:
        ...

class HasXDescriptor:
    x: Final[ReadOnlyDescriptor[int]] = ReadOnlyDescriptor()

class HasXAttr:
    x = 1

class HasXProperty:
    @property
    def x(self) -> int:
        ...

def f(arg: HasReadOnlyXProtocol[int]):
    print(arg.x + 1)

f(HasXDescriptor())
f(HasXAttr())
f(HasXProperty())
```

**Comment by draustin** [comment]
Thanks and sorry for the slow reply! That certainly meets my stated requirements, and also helped a lot with my understanding. Unfortunately it's unlikely to work for me for the following reasons.

1. The repetition would undermine the readability that I'm trying to achieve through static typing. 
2. My protocols typically have multiple properties/attributes/descriptors. As I understand it, to extend the method you describe to multiple properties would require every combination of descriptor or property to be called out as a protocol i.e. 2^n protocols required.

Interestingly but hypothetically (for now), I believe that a [type intersection operator](https://github.com/python/typing/issues/213) would address 2, but not 1. The ideal solution would be a way of [declaring that a protocol has an attribute, regardless of the implementation of that attribute](https://discuss.python.org/t/need-a-way-to-type-hint-attributes-that-is-compatible-with-duck-typing/41137/19) - which I realize is an issue for Python, not Pyright. Thanks again.
